integ/wait_until_search_attributes_optimization_test.go

```diff
@@ -0,0 +1,172 @@
+		WorkflowTimeoutSeconds: 15,
+		IwfWorkerUrl:           "http://localhost:" + testWorkflowServerPort,
+		StartStateId:           ptr.Any(wait_until_search_attributes_optimization.State1),
+		WorkflowStartOptions: &iwfidl.WorkflowStartOptions{
+			WorkflowConfigOverride: config,
+		},
+	}
+	_, httpResp, err := reqStart.WorkflowStartRequest(wfReq).Execute()
+	panicAtHttpError(err, httpResp)
+
+	time.Sleep(time.Second * 5)
+
+	signalValue := iwfidl.EncodedObject{
+		Encoding: iwfidl.PtrString("json"),
+		Data:     iwfidl.PtrString("test"),
+	}
+
+	reqSignal := apiClient.DefaultApi.ApiV1WorkflowSignalPost(context.Background())
+	httpResp, err = reqSignal.WorkflowSignalRequest(iwfidl.WorkflowSignalRequest{
+		WorkflowId:        wfId,
+		SignalChannelName: wait_until_search_attributes_optimization.SignalName,
+		SignalValue:       &signalValue,
+	}).Execute()
+
+	panicAtHttpError(err, httpResp)
+
+	reqWait := apiClient.DefaultApi.ApiV1WorkflowGetWithWaitPost(context.Background())
+	_, httpResp, err = reqWait.WorkflowGetRequest(iwfidl.WorkflowGetRequest{
+		WorkflowId: wfId,
+	}).Execute()
+	panicAtHttpError(err, httpResp)
+
+	// wait for workflow to complete
+	resp, httpResp, err := reqWait.WorkflowGetRequest(iwfidl.WorkflowGetRequest{
+		WorkflowId: wfId,
+	}).Execute()
+	panicAtHttpErrorOrWorkflowUncompleted(err, httpResp, resp)
+
+	api := uclient.GetApiService().(workflowservice.WorkflowServiceClient)
+	reqHistory := &workflowservice.GetWorkflowExecutionHistoryRequest{
+		Namespace: "default",
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wfId,
+		},
+	}
+	eventHistory, err := api.GetWorkflowExecutionHistory(context.Background(), reqHistory)
+	if err != nil {
+		panic("couldn't load eventHistory")
+	}
+
+	var upsertSAEvents []*history.HistoryEvent
+
+	for _, e := range eventHistory.History.Events {
+		if e.EventType == enums.EVENT_TYPE_UPSERT_WORKFLOW_SEARCH_ATTRIBUTES {
+			upsertSAEvents = append(upsertSAEvents, e)
+		}
+	}
+
+	switch mode := config.GetExecutingStateIdMode(); mode {
+	case iwfidl.ENABLED_FOR_ALL:
+		assertions.Equal(10, len(upsertSAEvents))
+		assertions.Equal([]string{"S1"}, historyEventSAs(upsertSAEvents[0]))
+		assertions.Equal([]string{"S2"}, historyEventSAs(upsertSAEvents[1]))
+		assertions.Equal([]string{"S2", "S3"}, historyEventSAs(upsertSAEvents[2]))
+		assertions.Equal([]string{"S3", "S4"}, historyEventSAs(upsertSAEvents[3]))
+		assertions.Equal([]string{"S3", "S5"}, historyEventSAs(upsertSAEvents[4]))
+		assertions.Equal([]string{"S3", "S6", "S7"}, historyEventSAs(upsertSAEvents[5]))
+		assertions.Equal([]string{"S3", "S6"}, historyEventSAs(upsertSAEvents[6]))
+		// TODO: This is unexpected; should not upsert the same SAs -- happens after "_SYS_GRACEFUL_COMPLETING_WORKFLOW"
+		assertions.Equal([]string{"S3"}, historyEventSAs(upsertSAEvents[7]))
+		assertions.Equal([]string{"S3"}, historyEventSAs(upsertSAEvents[8]))
+		assertions.Equal([]string{"null"}, historyEventSAs(upsertSAEvents[9]))
```

Comment:
Technically we should not upsert "S3" twice. Seems that we always upsert on `_SYS_GRACEFUL_COMPLETING_WORKFLOW`

- SAs -> "S3", "S6", "S7"
- S7-1 -> _SYS_GRACEFUL_COMPLETING_WORKFLOW
- SAs -> "S3", "S6"
- S6-1 -> _SYS_GRACEFUL_COMPLETING_WORKFLOW
- SAs -> "S3"
- S3-1 -> _SYS_GRACEFUL_COMPLETING_WORKFLOW
- SAs -> "S3"
- S3-2 -> _SYS_GRACEFUL_COMPLETING_WORKFLOW
- SAs -> null

Reply:
Is that because of we miss an existing optimization— on MarkCompletion, we should only refresh when any stateID reaches to zero (needs to delete)

Reply:
Okay, this should be the last thing to tackle. I'll take care of it in the morning

Reply:
Latest thought: If the NDE is actually because of sorting and we can fix it and then move the logic back to the refresh method, then this should not be an issue and we don’t need to check if a StateId count reaches zero

Reply:
Fixed this issue now and tests are passing as expected. We can try moving back to the refresh method